File: 500.600/507.PerfectNumber.cpp
#include <iostream>
#include <cmath>
using namespace std;

class PerfectNumber
{
public:
    bool checkPerfectNumber(int num)
    {
        if (num == 1)
        {
            return false;
        }
        int x = int(sqrt(num));
        int sum = 1;
        while (x > 1)
        {
            if (num % x == 0)
            {
                sum += x;
                sum += num / x;
            }
            --x;
        }
        return sum == num;
    }
};

int main(int argc, char const *argv[])
{
    PerfectNumber cls;
    cout << boolalpha << cls.checkPerfectNumber(1) << " -> true" << endl;
    cout << boolalpha << cls.checkPerfectNumber(28) << " -> true" << endl;
    cout << boolalpha << cls.checkPerfectNumber(7) << " -> false" << endl;
    return 0;
}

File: go/500.600/507.PerfectNumber.go
package main

import (
	"fmt"
	"math"
)

func checkPerfectNumber(num int) bool {
	if num == 1 {
		return false
	}
	x := int(math.Sqrt(float64(num)))
	sum := 1
	for x > 1 {
		if num%x == 0 {
			sum += x
			sum += num / x
		}
		x--
	}
	return sum == num
}

func main() {
	fmt.Println(checkPerfectNumber(28), true)
	fmt.Println(checkPerfectNumber(7), false)
}
